grails-app/services/org/pih/warehouse/inventory/StockMovementService.groovy

```diff
@@ -1083,12 +1087,109 @@ class StockMovementService {
+
+        data?.eachWithIndex { Map params, index ->
+            params.errors = new ArrayList()
+            if (!params.id) {
+                params.errors.push("Requisition item id is required")
+            }
+            if (params.quantity == null) {
+                params.errors.push("Quantity picked are required")
+            }
+            if (params.lot?.contains("E") && NumberUtils.isNumber(params.lot)) {
+                params.errors.push("Lot numbers must not be specified in scientific notation. " +
+                        "Please reformat field with Lot Number: \"${params.lot}\" to a number format")
+            }
+
+            PickPageItem pickPageItem = pickPageItems.find { it.requisitionItem?.id == params.id }
+            if (!pickPageItem) {
+                params.errors.push("Requisition item id: ${params.id} not found")
+            }
+            if (pickPageItem) {
+                AvailableItem availableItem = pickPageItem.availableItems?.find { item ->
+                    Boolean binLocationMatches = params.binLocation ? item.binLocation?.name == params.binLocation : !item.binLocation
+                    Boolean lotMatches = params.lot == item.inventoryItem?.lotNumber
+                    binLocationMatches && lotMatches
+                }
+
+                if (!availableItem) {
+                    params.errors.push("There is no item available with: ${params.lot ? "lot ${params.lot}" : ""} ${params.binLocation ? "bin ${params.binLocation}" : ""}")
+                }
+            }
+
+            Integer itemQuantitySum = groupedItems[params.id].sum { Integer.parseInt(it.quantity) }
+            if (pickPageItem && itemQuantitySum > pickPageItem.requisitionItem.quantity) {
+                if (!supportsOverPick) {
+                    params.errors.push("Item ${params.id} is overpicked, expected quantity ${pickPageItem.requisitionItem.quantity}, received ${itemQuantitySum}")
+                }
+            } else if (itemQuantitySum != pickPageItem.requisitionItem.quantity) {
+                params.errors.push("Item ${params.id} expected quantity is ${pickPageItem.requisitionItem.quantity}, received ${itemQuantitySum}")
+            }
+        }
+    }
+
+    void importPicklistTemplate(List<Map> data, StockMovement stockMovement, List<PickPageItem> pickPageItems) {
+        data.each { params ->
+            // skip rows with errors
+            if (params.errors) {
+                return
+            }
+
+            PickPageItem pickPageItem = pickPageItems.find {
+                it.requisitionItem?.id == params.id
+            }
+
+            AvailableItem availableItem = pickPageItem.availableItems?.find { item ->
+                Boolean binLocationMatches = params.binLocation ? item.binLocation?.name == params.binLocation : !item.binLocation
+                Boolean lotMatches = params.lot == item.inventoryItem?.lotNumber
+                binLocationMatches && lotMatches
```

Comment:
Since we're seeing this in a few places, let's go ahead and move this into the PickPageItem bean. 
```
AvailableItem availableItem = 
    pickPageItem.getAvailableItem(params.binLocation, params.lotNumber)
```

**PickPageItem**
```
AvailableItem getAvailableItem(Location binLocation, String lotNumber) {
    return availableItems?.find { item ->
                Boolean binLocationMatches = params.binLocation ? item.binLocation?.name == params.binLocation : !item.binLocation
                Boolean lotMatches = params.lot == item.inventoryItem?.lotNumber
                binLocationMatches && lotMatches
}
```
One question ... is there any way that we could have multiple available items matching on lot number and bin location. My gut says no, but just wanted to make sure.

Reply:
we shouln't. I dont think it is possible to create two stock records with same lot and binLocation